src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -228,9 +230,110 @@ public static char ToCharLower(int value)
+                // or some byte greater than 0x0f.
+                Vector128<byte> nibbles = Vector128.Min(t2 - Vector128.Create((byte)0xF0), t4);
+                // Any high bit is a sign that input is not a valid hex data
+                if (!Utf16Utility.AllCharsInVector128AreAscii(vec1 | vec2) ||
+                    AddSaturate(nibbles, Vector128.Create((byte)(127 - 15))).ExtractMostSignificantBits() != 0)
+                {
+                    // Input is either non-ASCII or invalid hex data
+                    break;
+                }
+                Vector128<byte> output;
+                if (Ssse3.IsSupported)
+                {
+                    output = Ssse3.MultiplyAddAdjacent(nibbles,
+                        Vector128.Create((short)0x0110).AsSByte()).AsByte();
+                }
+                else
+                {
+                    // Workaround for missing MultiplyAddAdjacent on ARM
+                    Vector128<short> ones = Vector128.Create(0x10010).AsInt16();
+                    Vector128<short> vl = AdvSimd.Multiply(
+                        AdvSimd.ZeroExtendWideningLower(nibbles.GetLower()).AsInt16(), ones).AsInt16();
+                    Vector128<short> vu = AdvSimd.Multiply(
+                        AdvSimd.ZeroExtendWideningLower(nibbles.GetUpper()).AsInt16(), ones).AsInt16();
+                    output = AdvSimd.AddSaturate(
+                        AdvSimd.Arm64.UnzipEven(vl, vu),
+                        AdvSimd.Arm64.UnzipOdd(vl, vu)).AsByte();
+                }
+                // Accumulate output in lower INT64 half and take care about endianness
+                output = Vector128.Shuffle(output, Vector128.Create((byte)0, 2, 4, 6, 8, 10, 12, 14, 0, 0, 0, 0, 0, 0, 0, 0));
+                // Store 8 bytes in dest by given offset
+                Unsafe.WriteUnaligned(ref Unsafe.Add(ref destRef, offset / 2), output.AsUInt64().ToScalar());
+
+                offset += (nuint)Vector128<ushort>.Count * 2;
+                if (offset == (nuint)chars.Length)
+                {
+                    return true;
+                }
+                // Overlap with the current chunk for trailing elements
+                if (offset > lengthSubTwoVector128)
+                {
+                    offset = lengthSubTwoVector128;
+                }
+            } while (true);
+
+            // Fall back to the scalar routine in case of invalid input.
+            return TryDecodeFromUtf16(chars.Slice((int)offset), bytes.Slice((int)offset / 2), out _);
+
+            // Remove once these are exposed as cross-plat helpers
+            static Vector128<byte> AddSaturate(Vector128<byte> a, Vector128<byte> b) =>
+                Sse2.IsSupported ? Sse2.AddSaturate(a, b) : AdvSimd.AddSaturate(a, b);
+            static Vector128<byte> SubtractSaturate(Vector128<byte> a, Vector128<byte> b) =>
+                Sse2.IsSupported ? Sse2.SubtractSaturate(a, b) : AdvSimd.SubtractSaturate(a, b);
```

Comment:
Can you make these internal APIs on Vector128? We've approved the API surface, just haven't exposed them yet.

Reply:
Done, thanks!